File: webauthn-server-core/src/main/scala/com/yubico/webauthn/impl/TokenBindingValidator.scala
package com.yubico.webauthn.impl

import java.util.Optional

import com.yubico.scala.util.JavaConverters._
import com.yubico.webauthn.data.TokenBindingInfo
import com.yubico.webauthn.data.TokenBindingStatus


object TokenBindingValidator {

  def validate(clientTokenBinding: Optional[TokenBindingInfo], rpTokenBindingId: Optional[String]): Boolean =
    rpTokenBindingId.asScala match {
      case None =>
        clientTokenBinding.asScala match {
          case None => true
          case Some(tbi) =>
            tbi.getStatus match {
              case TokenBindingStatus.SUPPORTED | TokenBindingStatus.NOT_SUPPORTED => true
              case TokenBindingStatus.PRESENT => throw new AssertionError("Token binding ID set by client but not by RP.")
            }
        }

      case Some(rpToken) =>
        clientTokenBinding.asScala match {
          case None => throw new AssertionError("Token binding ID set by RP but not by client.")
          case Some(tbi) =>
            tbi.getStatus match {
              case TokenBindingStatus.SUPPORTED | TokenBindingStatus.NOT_SUPPORTED => throw new AssertionError("Token binding ID set by RP but not by client.")
              case TokenBindingStatus.PRESENT =>
                tbi.getId.asScala match {
                  case None => throw new AssertionError("""Property "id" missing from "tokenBinding" object.""")
                  case Some(id) =>
                    assert(rpToken == id, "Incorrect token binding ID.")
                    true
                }
            }
        }
    }

}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/impl/TokenBindingValidator.java
package com.yubico.webauthn.impl;

import com.yubico.util.ByteArray;
import com.yubico.webauthn.data.TokenBindingInfo;
import java.util.Optional;


public class TokenBindingValidator {

    public static boolean validate(Optional<TokenBindingInfo> clientTokenBinding, Optional<ByteArray> rpTokenBindingId) {
        return rpTokenBindingId.map(rpToken ->
            clientTokenBinding.map(tbi -> {
                switch (tbi.getStatus()) {
                    case SUPPORTED:
                    case NOT_SUPPORTED:
                        throw new IllegalArgumentException("Token binding ID set by RP but not by client.");

                    case PRESENT:
                        return tbi.getId().map(id -> {
                            if (id.equals(rpToken)) {
                                return true;
                            } else {
                                throw new IllegalArgumentException("Incorrect token binding ID.");
                            }
                        }).orElseThrow(() -> new IllegalArgumentException("Property \"id\" missing from \"tokenBinding\" object."));
                }
                throw new RuntimeException("Unknown token binding status: " + tbi.getStatus());
            }).orElseThrow(() -> new IllegalArgumentException("Token binding ID set by RP but not by client."))
        ).orElseGet(() ->
            clientTokenBinding.map(tbi -> {
                switch (tbi.getStatus()) {
                    case SUPPORTED:
                    case NOT_SUPPORTED:
                        return true;

                    case PRESENT:
                        throw new IllegalArgumentException("Token binding ID set by client but not by RP.");
                }
                throw new RuntimeException("Unknown token binding status: " + tbi.getStatus());
            }).orElse(true)
        );
    }

}
